include/libcyphal/transport/can/media.hpp

```diff
@@ -77,21 +70,45 @@ class IMedia
     /// @param can_id The destination CAN ID of the frame.
     /// @return `true` if the frame is accepted or already timed out;
     ///         `false` to try again later (f.e. b/c output TX queue is currently full).
-    ///         If any media error occurred, the frame will be dropped by transport.
-    ///
-    virtual Expected<bool, MediaError> push(const TimePoint                    deadline,
-                                            const CanId                        can_id,
-                                            const cetl::span<const cetl::byte> payload) noexcept = 0;
+    ///         If any media failure occurred, the frame will be dropped by transport.
+    ///@{
+    struct PushResult
+    {
+        struct Success
+        {
+            bool is_accepted;
```

Comment:
I thought that I did it reasonably weighted... namely:
- use more simple return type approach for...
  - either non-fallible function, or
  - fallible but with `void` success (like f.e. `celt::optional<MediaFailure>`)
- use extra result type entity for more complex in nature returns

Take for example `struct ReceiveResult` for `IRxSocket::receive()` method. Yes, declaration site might look convoluted, but multiple usages of this declaration (implementation, mocks, tests) - they all operate with IMHO very well defined, isolated/nested visibility, and predictable! set of entities. They also don't have to "know" all the details, like f.e. know the exact list of all possible primitive error types inside of failure (like `cetl::variant<PlatformError, ArgumentError, MemoryError>`) - expanding or shrinking such set should be quite easy (if needed) b/c compilation will fail only where it matters. It will (were actually already several times) be harder to refactor code if I need **everywhere** type exact things (like `Expected<ReceiveRxMetadata, cetl::variant<PlatformError, ArgumentError, MemoryError>>`).

Reply:
Okay. I am not questioning the value of the interface segregation. An entity shall not depend on another entity unless it is needed, so of course we should only list the actual error types we need. I am just a bit concerned about the syntax, this is verbose:

```c++
struct PushResult
{
    struct Success
    {
        bool is_accepted;
    };
    using Failure = MediaFailure;

    using Type = Expected<Success, Failure>;
};
virtual PushResult::Type push...
```

This is not while functionally similar:

```c++
virtual Expected<bool, MediaFailure> push...
```

Reply:
... but instead of "cryptic" bool at `Expected<bool, ...` we have almost self-explanatory `bool is_accepted`.
Also, it's kinda "fortunate" that we have currently `MediaFailure`, and so `using Failure = MediaFailure;` statement looks like unnecessary, BUT I think we should not have `MediaFailure` in the first place, b/c different methods of the same `can::IMedia`, strictly speaking, might have different set of expected error types (maybe it's already the case btw, I need to look more closely). It's definitely the case for `udp::IMedia` methods.

What if we invest a bit into `Expected` instead (renaming it to `Result` template as well)?
So that `Success` & `Failure` type aliases will be automatically exposed from `Result`.
I'm not sure yet exactly how to do this, but my idea (of usage) will be like this:
```
using PushResult = Result<bool /*is_accepted*/, MediaFailure>;
virtual PushResult push(...) = 0;
```
and on call site it will be like:
```
auto push_result = media.push();
if (auto* const failure push_result.get_failure())  // hiding internal `cetl::get_if<Failure>(this)`
{
    ...
    return *failure;   // as optional<MediaFailure>, or `AnyFailure`
}
bool is_accepted = push_result.get_success();   // hiding internal `cetl::get<Success>(*this)`
```

Reply:
moved https://github.com/OpenCyphal-Garage/libcyphal/issues/369